source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1160,23 +1220,73 @@ enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
+  absl::string_view sni = absl::NullSafeStringView(
+      SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
 
-  // Fallback on first certificate.
-  const TlsContext* selected_ctx = &tls_contexts_[0];
-  auto ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
-  for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable != ctx.is_ecdsa_) {
-      continue;
-    }
+  const TlsContext* selected_ctx = nullptr;
+  OcspStapleAction ocsp_staple_action;
+  ServerNamesMap::iterator server_name_it;
 
+  auto selected = [&](const TlsContext& ctx) -> bool {
     auto action = ocspStapleAction(ctx, client_ocsp_capable);
     if (action == OcspStapleAction::Fail) {
-      continue;
+      // The selected ctx must adhere to OCSP policy
+      return false;
     }
 
-    selected_ctx = &ctx;
-    ocsp_staple_action = action;
-    break;
+    if (client_ecdsa_capable == ctx.is_ecdsa_ || selected_ctx == nullptr) {
+      // A cert might be selected as fallback even if the key type mismatches.
+      selected_ctx = &ctx;
+      ocsp_staple_action = action;
+    }
+
+    if (client_ecdsa_capable != ctx.is_ecdsa_) {
+      // It indicates trying next ctx and find a cert that matches key type.
+      return false;
+    }
+
+    return true;
+  };
+
+  auto select_from_map = [this, &selected](ServerNamesMap::iterator& server_name_it) -> void {
+    if (server_name_it == server_names_map_.end()) {
+      return;
+    }
+    const auto& pkey_types_map = server_name_it->second;
+    for (const auto& entry : pkey_types_map) {
+      if (selected(entry.second.get())) {
+        break;
+      }
+    }
+  };
+
+  // Select cert based on SNI if SNI exists.
+  if (!sni.empty()) {
+    // Match on exact server name, i.e. "www.example.com" for "www.example.com".
+    server_name_it = server_names_map_.find(sni);
+    select_from_map(server_name_it);
+
+    if (selected_ctx == nullptr) {
+      // Match on wildcard domain, i.e. ".example.com" for "www.example.com".
+      // https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
+      size_t pos = sni.find('.', 1);
+      if (pos < sni.size() - 1 && pos != std::string::npos) {
+        absl::string_view wildcard = sni.substr(pos);
+        server_name_it = server_names_map_.find(static_cast<std::string>(wildcard));
+      }
+      select_from_map(server_name_it);
+    }
+  }
+  // Select cert from tls context list if SNI does not exist or no proper cert matches to SNI.
+  if (selected_ctx == nullptr) {
+    // Fallback on first certificate.
+    selected_ctx = &tls_contexts_[0];
+    ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
+    for (const auto& ctx : tls_contexts_) {
```

Comment:
we can not avoid O(n) lookup unless we limit the numbers of lookup. Getting OCSP response at config-load time does not make sense since the client needs up-to-date information.

Reply:
Would it be sufficient to only check the first 2 contexts? In the old version (before this PR), only 2 were allowed, so that's all that would be checked. One RSA and one EC. If the matching key type has a failed OCSP state, then fail the connection. What do you think?

Reply:
actually I still prefer O(n) lookup though there is potential DoS attack, only checking the first 2 contexts is a tricky method, but if you think it's acceptable I can update. My thought is highlight the potential DoS issue in document to make users to carefully use large number of certificates. I think this feature will be often used in gateway but not sidecar, I'm not sure how many certificates will be in gateway, I guess it will not be too many,  what do you think about it?

Reply:
It might be ok, as long as we clearly document the behavior, eg if there are many certs, and the client can't find a pkey-and-ocsp match, it will check them all.

If you have time, could you try to benchmark how many certs would need to be in the list to equal the same CPU use as a TLS handshake? I'm guessing the number is quite large; that would be the point where a malicious client may be able to double the CPU use for establishing connections.

Reply:
I document the behavior with a highlight note.

I don't have time to do the benchmark for now. But I agree with that only quite large number of certs will have obviously negative effect on TLS handshake. We can keep the O(N) operation and users might give feedback in the future.